### PR TITLE
Add `Permutation(Group).permute_index` and refactor group-matrix `@`

### PR DESCRIPTION
**Pull request by attila-i-szabo**
This PR adds the ~`permute_index`~ `apply_to_id` method to both `Permutation` and `PermutationGroup`. They take an array of integers as input and return what they are mapped to by the permutation (or all permutations along the 0th axis). This is useful for generating all symmetry-equivalent configurations of a set of sites. For example, you can get every square of a (PBC) square lattice as
```python
graph = nk.graph.Square(6)
square = [0,1,7,6] # this is a square
squares = graph.translation_group().permute_index(square)
```

Incidentally, I also implemented the application of whole `PermutationGroup`s and `PointGroup`s to an `Array` more efficiently. Now, these are also added to the dispatch system around `@`. Everything works the same way as it used to, except now `group @ array` is valid and equivalent to `group(array)`.

### CODE REVIEW

File: netket/utils/group/_permutation_group.py [hunk 1]
```diff
@@ -199,6 +203,10 @@ def shape(self) -> Shape:
         """
         return (len(self), self.degree)
 
+    def permute_index(self, x: Array):
+        """Returns the image of indices `x` under all permutations"""
+        return np.argsort(self.to_array())[:, x]  # default argsort along last axis :)
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/utils/group/_permutation_group.py:208]: I'd vote for inverse in that case?

File: netket/utils/group/_permutation_group.py [hunk 2]
```diff
@@ -66,6 +66,10 @@ def __repr__(self):
     def __array__(self, dtype: DType = None):
         return np.asarray(self.permutation, dtype)
 
+    def permute_index(self, x: Array):
+        """Returns the image of indices `x` under the permutation"""
+        return np.argsort(self.permutation)[x]
```
Comment (PhilipVinc) [inline netket/utils/group/_permutation_group.py:71]: I'm not a fan of this name... If I am not mistaken, `indices` are called `id/ids` in the Lattice class for example in `Lattice.id_from_position` so I'd prefer we are consistent with calling those id.

Then, `permute`, while correct is very specific... we could actually call this `apply_to_id`?

File: netket/utils/group/_permutation_group.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/utils/group/_permutation_group.py]: the upside is that using inverse we could have a cheaper computational cost upon repeated applications.

File: netket/utils/group/_semigroup.py [hunk 3]
```diff
@@ -148,7 +148,7 @@ def __call__(self, initial):
         Apply all group elements to all entries of `initial` along the last axis.
         """
         initial = np.asarray(initial)
-        return np.array([np.apply_along_axis(elem, -1, initial) for elem in self.elems])
+        return product(self, initial)
```
Comment (PhilipVinc) [inline netket/utils/group/_semigroup.py:151]: nice

File: netket/utils/group/_permutation_group.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/utils/group/_permutation_group.py:71]: the tests are also complaining that this function is not tested on Permutation, but only on `PermutationGroup`. 
Can you add a test for this specific function?

**Review by PhilipVinc — COMMENTED** [review]
Thanks @attila-i-szabo , that's great except for those two nitpicks

File: netket/utils/group/_permutation_group.py (hunk 2 above)
Comment (attila-i-szabo) [reply netket/utils/group/_permutation_group.py]: > the tests are also complaining that this function is not tested on Permutation, but only on `PermutationGroup`. Can you add a test for this specific function?

Done. The test highlights that `Identity` (of course) doesn't have `apply_to_id`. Should we add an `Element` optional arg to `PermGroup.apply_to_id` as a convenience like it is here: https://github.com/netket/netket/blob/08986f3bc85cf3ddd059d21d6f540a510d4b834e/netket/utils/group/_point_group.py#L405-L413

Comment (PhilipVinc) [reply netket/utils/group/_permutation_group.py]: I don't understand what you are proposing...

Comment (attila-i-szabo) [reply netket/utils/group/_permutation_group.py]: Something like
```python
def permute_index(self, x: Array, g: Optional[Element] = None):
        """Returns the image of indices `x` under the permutation"""
    if g is None:
        return np.argsort(self.permutation)[x]
    else:
        if isinstance(g, Permutation):
            return ...
        elif isinstance(g, Identity):
            return ...
        else:
            raise ....
```
The point is that this way the identity element can be treated the same way as permutations.

Comment (attila-i-szabo) [reply netket/utils/group/_permutation_group.py]: I've just noticed this comment is on `Permutation.apply_to_id`. I meant `PemrutationGroup.apply_to_id`

**Review by PhilipVinc — APPROVED** [review]

**Pull request merged**